### PR TITLE
feat: formula for stake and unstake, and fetch data hook for calculation

### PR DESCRIPTION
**Pull request by raman-pelith**
clickup: CU-2ufku44

### CODE REVIEW

File: src/hooks/staking/useVaults.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    const providers = useMemo(() => {
+      const ethProvider = new JsonRpcProvider(process.env.REACT_APP_NETWORK_URL);
+      const bscProvider = new JsonRpcProvider(process.env.REACT_APP_BSC_NETWORK_URL);
+      const polygonProvider = new JsonRpcProvider(process.env.REACT_APP_POLYGON_NETWORK_URL);
+      const kovanProvider = new JsonRpcProvider(process.env.REACT_APP_KOVAN_NETWORK_URL);
+      return {[ChainId.MAINNET]: ethProvider, [ChainId.BSC]: bscProvider, [ChainId.POLYGON]: polygonProvider, [ChainId.KOVAN]: kovanProvider};
+    }, [])
+
+    const fetchVaults = async () => {
```
Comment (paul-pelith) [inline src/hooks/staking/useVaults.tsx:33]: should set chainId to the function
```
const fetchVaults = async (chainId: ChainId) => {}
```

then wrap this by using `useCallback`, then line: 48 can remove fetchVaults from the dependencies.
afterall, line 50 `useEffect` should update with new dependencies.

this way can make users more clarify which is a constants value or `react state`. and avoid `react closure` issue

File: src/hooks/staking/useVaults.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import { Contract as MulticallContract, Provider as MulticallProvider } from '@pelith/ethers-multicall';
+import { useWeb3React } from '@web3-react/core';
+import { JsonRpcProvider } from '@ethersproject/providers';
+import debounce from 'lodash.debounce';
+import { ChainDataFetchingState, NEW_SHAKKA_ADDRESSES, ChainId } from '../../constants';
+import STAKING_ABI from '../../constants/abis/shakka.json';
+import { useBlockNumber } from '../../state/application/hooks';
+
+export default function useVaults(): {
+  vaults: any;
```
Comment (paul-pelith) [inline src/hooks/staking/useVaults.tsx:11]: repeat in line 17

File: src/hooks/staking/useVaults.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  vaultCount: any;
```
Comment (paul-pelith) [inline src/hooks/staking/useVaults.tsx:12]: repeated in line 18

File: src/hooks/staking/useStakingRate.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import { useWeb3React } from '@web3-react/core';
+import { JsonRpcProvider } from '@ethersproject/providers';
+import { BigNumber } from '@ethersproject/bignumber';
+import debounce from 'lodash.debounce';
+import { ChainDataFetchingState, NEW_SHAKKA_ADDRESSES, ChainId } from '../../constants';
+import STAKING_ABI from '../../constants/abis/shakka.json';
+import { useBlockNumber } from '../../state/application/hooks';
+
+export default function useStakingRate(): {
+  stakingRate: BigNumber;
```
Comment (paul-pelith) [inline src/hooks/staking/useStakingRate.tsx:12]: repeated at line:17

File: src/hooks/staking/useVaults.tsx (hunk 2 above)
Comment (paul-pelith) [inline src/hooks/staking/useVaults.tsx:11]: should add return type

File: src/hooks/staking/useVaults.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  fetchDataState: ChainDataFetchingState;
+  } {
+    const { chainId } = useWeb3React();
+    const latestBlockNumber = useBlockNumber();
+    const [vaults, setVaults] = useState<any>();
+    const [vaultCount, setVaultCount] = useState<any>();
```
Comment (paul-pelith) [inline src/hooks/staking/useVaults.tsx:18]: type should be BigNumber

File: src/hooks/staking/useStakingRate.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  fetchDataState: ChainDataFetchingState;
+  } {
+    const { chainId } = useWeb3React();
+    const latestBlockNumber = useBlockNumber();
+    const [stakingRate, setStakingRate] = useState<BigNumber>();
```
Comment (paul-pelith) [inline src/hooks/staking/useStakingRate.tsx:17]: initialValue probably `BigNumber.from(0)`?

File: src/utils/stakeReceivedAmount.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  index: number,
+  amount: string, 
+  time: string, // the unit is year
+  stakingRate?: BigNumber,
+  vault?: any,
+): string[] | undefined[] {
+if (!stakingRate || !vault) { 
+  return [];
+}
+if (parseFloat(time) >  4 || parseFloat(time) < 30/60/24/360) {
```
Comment (hung-pelith) [inline src/utils/stakeReceivedAmount.ts:26]: Change 30/60/24/360 to constant

File: src/utils/stakeReceivedAmount.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+export function stakeReceivedAmount(
+    amount: string, 
+    time: string, // the unit is year
+    stakingRate?: BigNumber,
+  ): string | undefined {
+  if (!stakingRate) { 
+    return undefined;
+  }
+  const receivedSHakkaAmount = parseFloat(amount) * Math.pow(2, parseFloat(time)) * parseFloat(formatUnits(stakingRate, 18)) / 16;
```
Comment (hung-pelith) [inline src/utils/stakeReceivedAmount.ts:12]: Make receivedSHakkaAmount a function

File: src/utils/stakeReceivedAmount.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  stakingRate?: BigNumber,
+  vault?: any,
+): string[] | undefined[] {
+if (!stakingRate || !vault) { 
+  return [];
+}
+if (parseFloat(time) >  4 || parseFloat(time) < 30/60/24/360) {
+  return [];
+}
+const receivedSHakkaAmount =
```
Comment (hung-pelith) [inline src/utils/stakeReceivedAmount.ts:29]: And here

**Pull request merged**